lib/spack/spack/util/compression.py

```diff
@@ -278,6 +275,208 @@ def decompressor_for(path, ext):
+    def decomp_in_memory(cls, stream):
+        # python has no method of decompressing `.Z` files
+        return None
+
+
+class GZipFileType(CompressedFileType):
+    _MAGIC_NUMBER = b'\x1f\x8b\x08'
+    _ext = 'gz'
+
+    @classmethod
+    def decomp_in_memory(cls, stream):
+        if is_gzip_supported():
+            return gzip.GzipFile(fileobj=stream).read(cls.offset() + cls.header_size())
+        return None
+
+
+class LzmaFileType(CompressedFileType):
+    _MAGIC_NUMBER = b'\xfd7zXZ'
+    _ext = 'xz'
+
+    @classmethod
+    def decomp_in_memory(cls, stream):
+        if is_lzma_supported():
+            return lzma.LZMADecompressor().decompress(
+                stream, max_length=cls.offset() + cls.header_size())
+        return None
+
+
+class TarFileType(FileType):
+    _OFFSET = 257
+    _GNU_MAGIC = b'ustar  \0'
+    _POSIX_MAGIC = b'ustar\x0000'
+    _ext = 'tar'
+
+    @classmethod
+    def magic_number(cls):
+        return [cls._GNU_MAGIC, cls._POSIX_MAGIC]
+
+    @classmethod
+    def header_size(cls):
+        return max(len(cls._GNU_MAGIC), len(cls._POSIX_MAGIC))
+
+    @classmethod
+    def _bytes_check(cls, magic_bytes):
+        for magic in cls.magic_number():
+            if magic_bytes.startswith(magic):
+                return True
+        return False
+
+
+class ZipFleType(FileType):
+    _MAGIC_NUMBER = b'PK\003\004'
+    _ext = 'zip'
+
+
+VALID_FILETYPES = [BZipFileType, ZCompressedFileType,
+                   GZipFileType, LzmaFileType, TarFileType,
+                   ZipFleType]
+
+
+def extension_from_stream(stream, decompress=False):
+    """Return extension represented by stream corresponding to archive file
+    If stream does not represent an archive type recongized by Spack
+    (see `spack.util.compression.ALLOWED_ARCHIVE_TYPES`) method will return None
+
+    Extension type is derived by searching for identifying bytes
+    in file stream.
+
+    Args:
+        stream (IOBase): stream representing a file on system
+        decompress (bool) : if True, compressed files are checked
+                            for archive types beneath compression i.e. tar.gz
+                            default is False, otherwise, return top level type i.e. gz
+
+    Return:
+        extension (str|None): A string represting corresponding archive extension
+            or None as relevant.
+
+    """
+    for arc_type in VALID_FILETYPES:
+        if arc_type.is_file_of_type(stream):
+            suffix_ext = arc_type.extension()
+            prefix_ext = ''
+            if arc_type.is_compression_type() and decompress:
+                # stream represents compressed file
+                # get decompressed stream (if possible)
+                decomp_stream = arc_type.decomp_in_memory(stream)
+                if not decomp_stream:
+                    # we were unable to decompress, fallback on
+                    # filename parsing to derive extension
+                    return extension(stream.name)
```

Comment:
Now that we have two methods of extension detection, IMO it would be helpful to distinguish them more by name: i.e. "name_parsing_extension_detection" and "magic_number_extension_detection" (feel free to choose shorter names but FWIW you can make them as long as you wish and it won't bother me).